subprojects/core/src/integTest/groovy/org/gradle/api/tasks/IncrementalTasksIntegrationTest.groovy

```diff
@@ -108,4 +112,70 @@ class FailingIncrementalTasksIntegrationTest extends AbstractIntegrationSpec {
         "remove"      | false       | "with removed outputs is fully rebuilt"
         "none"        | true        | "with unmodified outputs is executed as incremental"
     }
+
+    @Issue("https://github.com/gradle/gradle/issues/9320")
+    def "incremental task with NAME_ONLY input (matching file names and content) detects changed input"() {
+        def taskName = "incrementalTask"
+        def inputs = ('a'..'z').collect { file("${it}/input.txt").createFile() }
+        def modifiedInput = inputs.toList().sort { random() }.first()
+
+        @Language("Groovy") script = """
+            class IncrementalTask extends DefaultTask {
+                @InputFiles
+                @PathSensitive(PathSensitivity.NAME_ONLY)
+                FileCollection inputFiles = project.files(${inputs.collect { "'${it}'" }.join(", ")})
+            
+                @TaskAction
+                def action(IncrementalTaskInputs inputs) {}
+            }
+            
+            task ${taskName}(type: IncrementalTask) {}
+        """
+        buildScript script
+        run taskName, "--info"
+
+        when:
+        modifiedInput.text = "changed"
+        println "Modified ${modifiedInput} to '${modifiedInput.text}'!"
```

Comment:
Instead of relying on `--info` output, check the assertion in the task action.

Reply:
That's a good point. I should have suggested it originally.

If we don't test the actual output, we need to make sure the task (and the assertions within) actually executed (i.e. add `executedAndNotSkipped(taskName)` to the test).

Reply:
I'll need some assistance here :)

Reply:
What @wolfs means is to add `assert` calls inside the task's action to ensure we got the right data via the `IncrementalTaskInputs` object.

Reply:
Addressed most of your concerns, thanks for taking the time.
@wolfs, @lptr, could you please push on top of this any change you would still like?
May be simpler than explanation.

Reply:
Not sure how to do that nicely to support both `IncrementalTaskInputs` and `InputChanges` ... will think of something

Reply:
I tried, but the source task was too different from the task that was posted in the issue.
That's why I added it here with parameters.
@wolfs said it's okay to keep it here for now.
Do you still want me to move it over?